### PR TITLE
/nodes POST and /nodes/<node_id> DELETE enabled

### PR DESCRIPTION
**Pull request by zzxgzgz**
/nodes POST and /nodes/<node_id> DELETE enabled.
now user can add node and delete node, and droplets for that node will be created/deleted.

### CODE REVIEW

File: src/mgmt/operator/obj/droplet.py [hunk 1]
```diff
@@ -36,9 +36,11 @@ def __init__(self, name, obj_api, opr_store, spec=None):
         super().__init__(name, obj_api, opr_store, spec)
         self.kind = "Droplet"
         self.plural = "droplets"
-        self.ip = ""
-        self.mac = ""
+        self.ip = []
+        self.mac = []
+        self.zgc_id = ''
         self.phy_itf = 'eth0'
```
Comment (liangbin-pub) [inline src/mgmt/operator/obj/droplet.py]: Can you change this to empty, eth0 is not a good initial value

File: test/yaml/droplet.yaml [hunk 2]
```diff
@@ -8,7 +8,9 @@ kind: Droplet
 metadata:
   name: d0
 spec:
-  mac: "0000:0000:0000:0001"
-  ip: "10.0.0.1"
+  mac: 
+  - "0000:0000:0000:0001"
+  ip: 
+  - "10.0.0.1"
```
Comment (liangbin-pub) [inline test/yaml/droplet.yaml:15]: If changed to array, give at least 2 entries

File: src/mgmt/services/manager/requirements.txt [hunk 3]
```diff
@@ -4,3 +4,4 @@ flask-migrate==2.5.2
 Flask-SQLAlchemy==2.4.1
 gunicorn==20.0.4
 psycopg2-binary==2.8.4
+kubernetes==11.0.0
```
Comment (liangbin-pub) [inline src/mgmt/services/manager/requirements.txt]: Does this version needs to be fixed? If it is, please check compatibility with microk8s we used for non-kind deployment scenarios

File: src/mgmt/services/manager/project/api/nodes.py [hunk 4]
```diff
@@ -9,26 +9,181 @@
...
 )
-
+from functools import reduce
+from kubernetes.client.rest import ApiException
 from project.api.models import Node
 from project import db
+from .models import Zgc
+from kubernetes import client, config
+import json
+from .vpcs import getGWsFromIpRange, MAC_PRIVATE, int_to_mac
```
Comment (liangbin-pub) [inline src/mgmt/services/manager/project/api/nodes.py]: suggest make a utility module for these small utilities

File: src/mgmt/services/manager/project/api/nodes.py [hunk 5]
```diff
@@ -9,26 +9,181 @@
...
+
+        # Try to get the existing droplets        
+        all_droplets_in_zgc = obj_api.list_cluster_custom_object(group='zeta.com',
+                                                                 version='v1',
+                                                                 plural='droplets',
+                                                                 label_selector='zgc_id='+post_data['zgc_id']+',network=tenant'
+                                                                )['items']
+        zgc_ip_list = post_data['ip_control'].split('.')
+        zgc_ip_list[0] = '10'
+        zgc_ip_list[1] = '0'
```
Comment (liangbin-pub) [inline src/mgmt/services/manager/project/api/nodes.py]: The CIDR for zgc_ip should coming from a configmap and also make sure zgc network created for KIND in create-cluster.sh matches, otherwise FTNs can't reach each other

File: src/mgmt/services/manager/project/api/nodes.py [hunk 6]
```diff
@@ -9,26 +9,181 @@
...
+                        droplet_ip_list = droplet_spec['ip']
+                        if len(droplet_ip_list) > 1:
+                            popped_ip = droplet_ip_list.pop()
+                            ip_for_new_droplet.append(popped_ip)
+                            droplet_mac_list = droplet_spec['mac']
+                            mac_from_ip =  get_mac_from_ip(popped_ip)
+                            if mac_from_ip in droplet_mac_list:
+                                droplet_mac_list.remove(mac_from_ip)
+                            else:
+                                droplet_mac_list.pop()
```
Comment (liangbin-pub) [inline src/mgmt/services/manager/project/api/nodes.py:145]: Store IP/Mac in separate array is not ideal exactly the issue you are trying to avoid here. Suggest an array of dict with ip and mac as fields, so they append/pop together

File: src/mgmt/services/manager/project/api/nodes.py [hunk 7]
```diff
@@ -9,26 +9,181 @@
...
+                    if current_length_of_ip_list == len(ip_for_new_droplet):
+                        logger.info('Breaking the while loop as there is no new ip added')
+                        break
+                    else:
+                        current_length_of_ip_list = len(ip_for_new_droplet)
+                macs_for_new_droplet = [get_mac_from_ip(ip) for ip in ip_for_new_droplet]
+
+                for droplet in droplet_that_can_give_ip_mac:
+                    update_droplet(droplet)
+
```
Comment (liangbin-pub) [inline src/mgmt/services/manager/project/api/nodes.py:157]: First we need to update only droplets actually changed.
I have a question: can we make sure the all_droplets_in_zgc list is always in same order? otherwise the list may not be sorted by number of IPs and when you iterate through it you need to make sure not giving up IP for droplet if its ip[] len <= number_of_ip_new_droplet_gets (example 8 IPs to (node1,node2,node3) => delete node 1 => add node 1 =>add node 4 you may ended up as node1 (1) node2(2) node3(3) node4(2).

File: src/mgmt/services/manager/project/api/nodes.py [hunk 8]
```diff
@@ -9,26 +9,181 @@
...
+                create_droplet(name='droplet-' +'tenant-' + post_data['name'].replace('_', "-").lower(), 
+                               ip=ip_for_new_droplet,
```
Comment (liangbin-pub) [inline src/mgmt/services/manager/project/api/nodes.py:159]: 'droplet-' +'tenant-' => 'droplet-tenant-'

File: src/mgmt/services/manager/project/api/nodes.py [hunk 9]
```diff
@@ -9,26 +9,181 @@
...
+                               zgc_id=post_data['zgc_id'])
+                create_droplet(name='droplet-' +'zgc-' + post_data['name'].replace('_', "-").lower(), 
+                               ip=[zgc_ip], 
+                               mac=[zgc_mac],
+                               itf=post_data['inf_zgc'], 
+                               network='zgc', 
+                               zgc_id=post_data['zgc_id'])
+            else:
+                logger.info("There's no zgc with zgc_id: {} in the database!".format(post_data['zgc_id']))
+        # commit change to data at last
```
Comment (liangbin-pub) [inline src/mgmt/services/manager/project/api/nodes.py:195]: This should be an error and respond POST as invalid parameter

File: src/mgmt/services/manager/project/api/nodes.py [hunk 10]
```diff
@@ -9,26 +9,181 @@
...
         db.session.add(Node(**post_data))
         db.session.commit()
-        response_object = post_data
```
Comment (liangbin-pub) [inline src/mgmt/services/manager/project/api/nodes.py:31]: why this change?

File: src/mgmt/services/manager/project/api/nodes.py [hunk 11]
```diff
@@ -62,6 +217,50 @@ def single_node(node_id):
...
+
+        for droplet in all_droplets_in_zgc:
+            if droplet['metadata']['name'] == tenant_droplet_name:
+                droplet_to_remove = droplet
+                all_droplets_in_zgc.remove(droplet)
+                break
+
+        if droplet_to_remove is None:
+            logger.error('Cannot find that droplet with name: {}, thus not deleting anything'.format(tenant_droplet_name))
+            return jsonify({})
```
Comment (liangbin-pub) [inline src/mgmt/services/manager/project/api/nodes.py]: REST respond with internal error

File: src/mgmt/services/manager/project/api/nodes.py [hunk 12]
```diff
@@ -62,6 +217,50 @@ def single_node(node_id):
...
+            ip_amount = len(ip_to_assign)
+
+            number_of_droplets_to_assign = len(all_droplets_in_zgc)
+
+            modified_droplets = []
+            for index in range(ip_amount):
+                ip = ip_to_assign[index]
+                mac = get_mac_from_ip(ip)
+                all_droplets_in_zgc[index % number_of_droplets_to_assign]['spec']['ip'].append(ip)
+                all_droplets_in_zgc[index % number_of_droplets_to_assign]['spec']['mac'].append(mac)
```
Comment (liangbin-pub) [inline src/mgmt/services/manager/project/api/nodes.py:266]: This algorithm and previous give up logic both have issue of unbalanced distribution. Consider sth like giving up in descending order and distributing in ascending order (sorted len(ip))

File: src/mgmt/services/manager/project/api/nodes.py [hunk 13]
```diff
@@ -62,6 +217,50 @@ def single_node(node_id):
...
+                if all_droplets_in_zgc[index % number_of_droplets_to_assign] not in modified_droplets: 
+                    modified_droplets.append(all_droplets_in_zgc[index % number_of_droplets_to_assign])
+
```
Comment (liangbin-pub) [inline src/mgmt/services/manager/project/api/nodes.py:270]: dictionary may be better than list to avoid checking existence

**Review by liangbin-pub — COMMENTED** [review]
All logic looks good, some comments to consider

File: src/mgmt/operator/obj/droplet.py (hunk 1 above)
Comment (zzxgzgz) [reply src/mgmt/operator/obj/droplet.py]: Will do.

File: src/mgmt/services/manager/project/api/nodes.py (hunk 13 above)
Comment (zzxgzgz) [reply src/mgmt/services/manager/project/api/nodes.py:270]: You're right, that was also my initial implementation and it didn't work, it is because that those droplets are already dictionaries, and in Python we cannot put dictionaries in another dictionary.

File: src/mgmt/services/manager/requirements.txt (hunk 3 above)
Comment (zzxgzgz) [reply src/mgmt/services/manager/requirements.txt]: I checked for it, right now the python k8s version 17.0 supports K8s version 1.17. We're currently using k8s python api 11.0.0, and it supports K8s version 1.15. 
[Python K8s Compatibility](https://github.com/kubernetes-client/python/tree/master#compatibility)

For MicroK8s, the latest version supports K8s version 1.19, and its version 1.15 supports K8s version 1.15. 
[MicroK8s V 1.15](https://github.com/ubuntu/microk8s/blob/1.15/README.md)

We should definitely keep the versions synced, I think for now we can use MiCroK8s v 1.15, unless there's need for a high version K8s.

File: src/mgmt/services/manager/project/api/nodes.py (hunk 13 above)
Comment (liangbin-pub) [reply src/mgmt/services/manager/project/api/nodes.py:270]: https://www.programiz.com/python-programming/nested-dictionary

File: src/mgmt/services/manager/project/api/nodes.py (hunk 10 above)
Comment (zzxgzgz) [reply src/mgmt/services/manager/project/api/nodes.py:31]: I forgot why, but it is not expected behavior, will change it back.

File: src/mgmt/services/manager/project/api/nodes.py (hunk 6 above)
Comment (zzxgzgz) [reply src/mgmt/services/manager/project/api/nodes.py:145]: But here I'm trying to remove the IP and MAC from the existing ones, the IP and Mac are separate properties of a droplet.

The `mac_from_ip` already makes sure that I'm generating the expected mac to remove from the ip, the else statement here is only for the case when there's no that `mac_from_ip`, which should never happen.

**Comment by liangbin-pub** [comment]
This PR is replaced by PR#45
